File: java/L100/LeetCode0107.java
package L100;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import util.TreeNode;

public class LeetCode0107 {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        if (root == null) {
            return res;
        }
        Queue<TreeNode> qu = new LinkedList<>();
        qu.offer(root);
        while (!qu.isEmpty()) {
            int size = qu.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0;i < size;++i) {
                TreeNode cur = qu.poll();
                level.add(cur.val);
                if (cur.left != null) {
                    qu.offer(cur.left);
                }
                if (cur.right != null) {
                    qu.offer(cur.right);
                }
            }
            res.add(0, level);
        }
        return res;
    }
}

File: cpp/L100/LeetCode0107.cc
#include "../heads/tree_node.h"
#include <algorithm>
#include <queue>
#include <vector>
using std::vector;

class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        std::queue<TreeNode*> q;
        vector<vector<int>> res;
        if (root == nullptr) {
            return res;
        }
        q.push(root);
        while (!q.empty()) {
            int qz = q.size();
            vector<int> level;
            for (int i = 0;i < qz;++i) {
                TreeNode *node = q.front();
                q.pop();
                level.push_back(node->val);
                if (node->left != nullptr) {
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }
            res.emplace_back(level);
        }
        std::reverse(res.begin(), res.end());
        return res;
    }
};